patches/tock/05-usb-cancel.patch

```diff
@@ -0,0 +1,747 @@
+             8 => unimplemented!("isochronous endpoint"),
+@@ -1405,25 +1388,25 @@ impl<'a> Usbd<'a> {
+             1..=7 => {
+                 // Notify the client about the new packet.
+                 let packet_bytes = regs.size_epout[endpoint].get();
+-                let (transfer_type, direction, state) =
++                let (transfer_type, in_state, out_state) =
+                     self.descriptors[endpoint].state.get().bulk_state();
+-                assert_eq!(state, BulkState::OutDma);
++                assert_eq!(out_state, Some(BulkOutState::OutDma));
+ 
+-                self.debug_packet("out", packet_bytes as usize, endpoint);
++                self.debug_out_packet(packet_bytes as usize, endpoint);
+ 
+                 self.client.map(|client| {
+                     let result = client.packet_out(transfer_type, endpoint, packet_bytes);
+                     debug_packets!("packet_out => {:?}", result);
+-                    let newstate = match result {
++                    let new_out_state = match result {
+                         hil::usb::OutResult::Ok => {
+                             // Indicate that the endpoint is ready to receive data again.
+                             regs.size_epout[endpoint].set(0);
+-                            BulkState::Init
++                            BulkOutState::Init
+                         }
+ 
+                         hil::usb::OutResult::Delay => {
+                             // We can't send the packet now. Wait for a resume_out call from the client.
+-                            BulkState::OutDelay
++                            BulkOutState::OutDelay
+                         }
+ 
+                         hil::usb::OutResult::Error => {
+@@ -1432,13 +1415,13 @@ impl<'a> Usbd<'a> {
+                                     + EndpointStall::IO::Out
+                                     + EndpointStall::STALL::Stall,
+                             );
+-                            BulkState::Init
++                            BulkOutState::Init
+                         }
+                     };
+                     self.descriptors[endpoint].state.set(EndpointState::Bulk(
+                         transfer_type,
+-                        direction,
+-                        newstate,
++                        in_state,
++                        Some(new_out_state),
+                     ));
+                 });
+             }
+@@ -1497,29 +1480,27 @@ impl<'a> Usbd<'a> {
+         // Endpoint 8 (isochronous) doesn't receive any EPDATA event.
+         for endpoint in 1..NUM_ENDPOINTS {
+             if epdatastatus.is_set(status_epin(endpoint)) {
+-                let (transfer_type, direction, state) =
++                let (transfer_type, in_state, out_state) =
+                     self.descriptors[endpoint].state.get().bulk_state();
+-                match state {
+-                    BulkState::InData => {
++                assert!(in_state.is_some());
```

Comment:
There are many assert/panic in this code. Are they only for internal invariants? Or also for preconditions (if this API is callable by some user)?